### PR TITLE
Add `NeighborLoader` benchmark suite

### PR DESCRIPTION
**Pull request by mszarma**
* Adding loader folder to benchmarks
*  Adding NeighborLoader benchmark for sampling on ognb-arxiv, ogbn-products and ogbn-mag

### CODE REVIEW

File: benchmark/sampling/neighborloader_bench.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import argparse
+import os.path as osp
+from timeit import default_timer
```
Comment (rusty1s) [inline benchmark/sampling/neighborloader_bench.py:3]: Would be cool to add/use some functionality from `torch_geometric.profile` for this, see [here](https://pytorch-geometric.readthedocs.io/en/latest/modules/profile.html). For example, `torch_geometric.profile.timeit`.

File: benchmark/sampling/neighborloader_bench.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+from torch_geometric.loader import NeighborLoader
+
+
+def run(args: argparse.ArgumentParser) -> None:
+
+    print("BENCHMARK STARTS")
+    for dataset_name in args.datasets:
+        print("Dataset: ", dataset_name)
+
+        if args.datasets_root == 'data':
```
Comment (rusty1s) [inline benchmark/sampling/neighborloader_bench.py:18]: Shouldn't we set `args.datasets_root` to `../../data` by default? Moving this two layers up looks a bit unexpected to me. We can also just leave in `root_dir = "data"` though.

File: benchmark/sampling/neighborloader_bench.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                            dataset_name.partition("-")[2])
+        else:
+            root = args.datasets_root
+
+        if dataset_name == 'ogbn-mag':
+            transform = T.ToUndirected(merge=True)
+            dataset = OGB_MAG(root=root, transform=transform)
+            train_idx = ('paper', dataset[0]['paper'].train_mask)
+            valid_idx = ('paper', dataset[0]['paper'].val_mask)
+            neighbour_sizes = args.hetero_neighbour_sizes
```
Comment (rusty1s) [inline benchmark/sampling/neighborloader_bench.py:30]:
```suggestion
            neighbor_sizes = args.hetero_neighbor_sizes
```

File: benchmark/sampling/neighborloader_bench.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+        print('Train sampling')
+        for sizes in neighbour_sizes:
+            print(f'Sizes={sizes}')
+            for batch_size in args.batch_sizes:
+                train_loader = NeighborLoader(data, num_neighbors=sizes,
+                                              input_nodes=train_idx,
+                                              batch_size=batch_size,
+                                              shuffle=True,
+                                              num_workers=args.num_workers)
```
Comment (rusty1s) [inline benchmark/sampling/neighborloader_bench.py:48]:
```suggestion
                                              num_workers=args.num_workers, )
```

File: benchmark/sampling/neighborloader_bench.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                    times.append(round(stop - start, 3))
+                average_time = round(sum(times) / args.runs, 3)
+                print(f'Batch size={batch_size}   iterations={iter}   ' +
+                      f'times={times}    average_time={average_time}')
+        print('Validation sampling')
+        for batch_size in args.eval_batch_sizes:
+            val_loader = NeighborLoader(data, num_neighbors=[-1],
+                                        input_nodes=valid_idx,
+                                        batch_size=batch_size, shuffle=False,
+                                        num_workers=args.num_workers)
```
Comment (rusty1s) [inline benchmark/sampling/neighborloader_bench.py:66]:
```suggestion
                                        num_workers=args.num_workers, )
```

File: benchmark/sampling/neighborloader_bench.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                    for batch in train_loader:
+                        iter = iter + 1
+                    stop = default_timer()
+                    times.append(round(stop - start, 3))
+                average_time = round(sum(times) / args.runs, 3)
+                print(f'Batch size={batch_size}   iterations={iter}   ' +
+                      f'times={times}    average_time={average_time}')
+        print('Validation sampling')
+        for batch_size in args.eval_batch_sizes:
+            val_loader = NeighborLoader(data, num_neighbors=[-1],
```
Comment (rusty1s) [inline benchmark/sampling/neighborloader_bench.py:63]: Add some comment why this benchmark is needed/why it relates to validation sampling.

**Review by rusty1s — APPROVED** [review]
Thank you! Left a few minor comments.

File: benchmark/sampling/neighborloader_bench.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
+import argparse
```
[the remark this answers is not in the document]
Comment (mszarma) [reply benchmark/sampling/neighborloader_bench.py]: Thanks Matthias for review, added new version. Done.

File: benchmark/sampling/neighborloader_bench.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+
+if __name__ == '__main__':
+    argparser = argparse.ArgumentParser('NeighborLoader Sampling Benchmarking')
+
+    argparser.add_argument('--device', default='cpu', type=str)
+    argparser.add_argument('--datasets', nargs="+",
+                           default=['ogbn-arxiv', 'ogbn-products',
+                                    'ogbn-mag'], type=str)
+    argparser.add_argument('--datasets-root', default='data', type=str)
```
[the remark this answers is not in the document]
Comment (mszarma) [reply benchmark/sampling/neighborloader_bench.py]: Done

File: benchmark/sampling/neighborloader_bench.py (hunk 1 above)
Comment (mszarma) [reply benchmark/sampling/neighborloader_bench.py]: How about adding more profiling in next PRs? and put this PR right now as a base.

File: benchmark/sampling/neighborloader_bench.py (hunk 2 above)
Comment (mszarma) [reply benchmark/sampling/neighborloader_bench.py]: Done - i left only root with "../../data" to not overcomplicate it. Other paths can be handle by redirecting by root.

File: benchmark/sampling/neighborloader_bench.py (hunk 3 above)
Comment (mszarma) [reply benchmark/sampling/neighborloader_bench.py]: Done

File: benchmark/sampling/neighborloader_bench.py (hunk 4 above)
Comment (mszarma) [reply benchmark/sampling/neighborloader_bench.py]: Done

File: benchmark/sampling/neighborloader_bench.py (hunk 5 above)
Comment (mszarma) [reply benchmark/sampling/neighborloader_bench.py]: Done

File: benchmark/sampling/neighborloader_bench.py (hunk 6 above)
Comment (mszarma) [reply benchmark/sampling/neighborloader_bench.py]: Initially i thought about checking also some scenario with sampling all neighbors like more inference specific sampling? i changed the code - let me know if that make sense or there are some better real scenarios? WDYT?

File: benchmark/sampling/neighborloader_bench.py (hunk 1 above)
Comment (rusty1s) [reply benchmark/sampling/neighborloader_bench.py]: Works for me.

**Pull request merged**